cmd/nanotube/main.go

```diff
@@ -75,21 +94,41 @@ func main() {
 
 	// SIGTERM gracefully terminates with timeout
 	// SIGKILL terminates immediately
+	// SIGUSR2 triggers zero-downtime restart
 	sgn := make(chan os.Signal, 1)
-	signal.Notify(sgn, os.Interrupt, syscall.SIGTERM)
-	<-sgn
+	signal.Notify(sgn, os.Interrupt, syscall.SIGTERM, syscall.SIGUSR2)
 
-	// start termination sequence
-	close(stop)
+	for {
+		s := <-sgn
 
-	select {
-	case <-time.After(time.Second * time.Duration(cfg.TermTimeoutSec)):
-		err = lg.Sync()
-		if err != nil {
-			fmt.Fprintf(os.Stderr, "error when syncing logger: %v", err)
+		if s == syscall.SIGUSR2 {
+			lg.Info("Reload: Got signal for reload. Checking config.")
+			_, _, _, _, _, err = loadBuildRegister(cfgPath, lg)
```

Comment:
Would it work without this check?
I'm thinking that on invalid config, StartProcess migth fail.

Reply:
This is part of the solution.

`StartProcess` won't fail; it will start the process and return a PID. Then the process will exit with non-zero return code, but we will never know and by then we will start termination.

This config check is something we very much want to prevent invalid config logs. It's the very thing we want.